Mathlib/Tactic/Linter/UnusedAssumptionInType.lean

```diff
@@ -0,0 +1,115 @@
+/-
+Copyright (c) 2024 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov, Michael Rothgang
+-/
+import Batteries.Tactic.Lint
+import Mathlib.Lean.Expr.Basic
+
+/-!
+# Linters for unused assumptions in a type
+
+This file defines a collection of similar linters, which check for declarations assuming a
+hypothesis of the form `[Type p]` which contains an assumption about `p` that is not used:
+this includes `Decidable p`, `Encodable p`, `Inhabited p` and others.
+Usually, the code can be restructured to avoid these assumptions, making the statements in question
+more general.
+-/
+
+namespace Batteries.Tactic.Lint
+open Lean Meta
+
+/-- Check if a given declaration assumes some hypothesis `[Type p]`, but doesn't use this
+assumption in the type. `typesToAvoid` the list of such types.
+This is the main logic underlying the linters below. -/
+def checkUnusedAssumptionInType (declInfo : ConstantInfo) (typesToAvoid : Array Name) :
+    MetaM (Option MessageData) := do
+  -- We omit inductive types and their constructors, to reduce false positives.
+  -- We also omit partial declarations (via the `opaque` definitions they generate):
+  -- these are less useful for theorem proving, hence the linter is less useful there.
+  if declInfo.isInductive || declInfo.isCtor ||
+      (declInfo matches .opaqueInfo ..) then return none
+  let type := declInfo.type
+  -- Compute an array of pairs (argument index, error message) for each superfluous argument:
+  -- the first component is the index of the superfluous argument, the second component
+  -- contains details about the error.
+  let mut impossibleArgs ← forallTelescopeReducing type fun args ty ↦ do
+    let argTys ← args.mapM inferType
+    let ty ← ty.eraseProofs
+    return ← (args.zip argTys.zipWithIndex).filterMapM fun (arg, t, i) ↦ do
+      unless typesToAvoid.any t.cleanupAnnotations.getForallBody.isAppOf do return none
+      let fv := arg.fvarId!
+      if ty.containsFVar fv then return none
+      if argTys[i+1:].any (·.containsFVar fv) then return none
+      return some (i, (← addMessageContextFull m!"argument {i+1} {arg} : {t}"))
+  if !(← isProp type) then
+    if let some e := declInfo.value? then
+      impossibleArgs ← lambdaTelescope e fun args e ↦ do
+        let e ← e.eraseProofs
+        return impossibleArgs.filter fun (k, _) ↦
+          k < args.size && !e.containsFVar args[k]!.fvarId!
+  if impossibleArgs.isEmpty then return none
+  return some <| .joinSep (impossibleArgs.toList.map Prod.snd) ", "
```

Comment:
I saw you were having performance problems. I tried rewriting this as efficiently as I could, but unfortunately I didn't test it, so it might be worthless... Here:
```lean
/-- Check if a given declaration assumes some hypothesis `[Type p]`, but doesn't use this
assumption in the type. `typesToAvoid` the list of such types.
This is the main logic underlying the linters below. -/
def checkUnusedAssumptionInType (declInfo : ConstantInfo) (typesToAvoid : Array Name) :
    MetaM (Option MessageData) := do
  -- We omit inductive types and their constructors, to reduce false positives.
  -- We also omit partial declarations (via the `opaque` definitions they generate):
  -- these are less useful for theorem proving, hence the linter is less useful there.
  if declInfo.isInductive || declInfo.isCtor ||
      (declInfo matches .opaqueInfo ..) then return none
  let type := declInfo.type
  -- Early return: none of the constants to avoid appear.
  unless type.containsConst typesToAvoid.contains do
    return none
  forallTelescopeReducing type fun args ty ↦ do
    -- Compute whether the given argument is used by the body of the definition outside of proofs.
    let usedIdxs : Array Bool ← do
      if ← pure declInfo.hasValue <&&> not <$> isProp type then
        lambdaTelescope declInfo.value! fun args' e ↦ do
          let mut usedIdxs := args.map (fun _ => false)
          let e ← e.eraseProofs
          let st := collectFVars {} e
          for i in [0:args'.size] do
            usedIdxs := usedIdxs.set! i (st.fvarSet.contains args'[i]!.fvarId!)
          pure usedIdxs
      else
        pure #[]
    -- Early return: every argument is used.
    if usedIdxs.size == args.size && usedIdxs.all id then
      return none
    -- Compute an array of error messages for each superfluous argument
    let ty ← ty.eraseProofs
    let mut impossibleArgs := #[]
    let mut usedFVars : CollectFVars.State := collectFVars {} ty
    for i' in [0:args.size] do
      let i := args.size - i' - 1
      let arg := args[i]!
      let t := (← inferType arg).cleanupAnnotations
      if t.containsConst typesToAvoid.contains then
        if usedIdxs[i]? != some true && !usedFVars.fvarSet.contains t.fvarId! then
          impossibleArgs := impossibleArgs.push
            (← addMessageContextFull m!"argument {i+1} {arg} : {t}")
      usedFVars := collectFVars usedFVars t
    if impossibleArgs.isEmpty then return none
    return some <| .joinSep impossibleArgs.reverse.toList ", "
```

Reply:
If written correctly, it should be nearly functionally identical. The only intentional difference is that the message uses `t` with cleaned-up annotations now.

Reply:
Thanks! I have used this version now; let's see how it fares w.r.t. speed!